general/src/store/normalize.rs

```diff
@@ -232,28 +232,37 @@ impl<'a> NormalizeProcessor<'a> {
 
     /// Inserts the IP address into the user interface. Creates the user if it doesn't exist.
     fn normalize_user_ip(&self, event: &mut Event) {
-        let remote_addr = event
+        // If there is no User ip_address, update it either from the Http interface
+        // or the client_ip of the request.
+        let ip_address = event
             .request
             .value()
             .and_then(|request| request.env.value())
             .and_then(|env| env.get("REMOTE_ADDR"))
             .and_then(Annotated::<Value>::as_str)
-            .and_then(|ip| IpAddr::parse(ip).ok());
-
-        // If there is no User ip_address, update it either from the Http interface
-        // or the client_ip of the request.
-        if let Some(ip_address) = remote_addr {
-            self.set_user_ip(&mut event.user, ip_address);
-        } else if let Some(ref ip_address) = self.config.client_ip {
-            self.set_user_ip(&mut event.user, ip_address.clone());
-        }
-    }
-
-    /// Creates or updates the user's IP address.
-    fn set_user_ip(&self, user: &mut Annotated<User>, ip_address: IpAddr) {
-        let user = user.value_mut().get_or_insert_with(User::default);
-        if user.ip_address.value().is_none() {
-            user.ip_address = Annotated::new(ip_address);
+            .and_then(|ip| IpAddr::parse(ip).ok())
+            .or_else(|| self.config.client_ip.clone());
+
+        if let Some(ip_address) = ip_address {
+            let platform = event.platform.value().map(String::as_str);
+            let user = event.user.value_mut().get_or_insert_with(User::default);
+
+            let is_auto = user
+                .ip_address
+                .value()
+                .map(|x| x.is_auto())
+                .unwrap_or(false);
+            let is_missing = user.ip_address.value().is_none();
+
+            // In an ideal world all SDKs would set {{auto}} explicitly.
+            let set_if_missing = match platform {
+                Some("javascript") | Some("cocoa") | Some("objc") => true,
+                _ => false,
+            };
```

Comment:
_nit_: This block can be folded to avoid the redundant check for an existing ip address:

```rust
let override_ip = match user.ip_address.value() {
    Some(ip_addr) => ip_addr.is_auto(),
    // In an ideal world all SDKs would set {{auto}} explicitly.
    None => match platform {
        Some("javascript") | Some("cocoa") | Some("objc") => true,
        _ => false,
    }
};
```